### ISSUE TITLE
Prepare installation device with NTFS as target filesystem took too long

### ISSUE THREAD

**Issue by sevku** [slacka/WoeUSB#247]
There should be a progress indication for the last step where it says `i386-pc installation`. In my case this took a full 50 minutes.

**Comment by Lin-Buo-Ren** [comment]
If you specify the `--verbose` option it should print out additional processing message, please check it out if it solve your problem.

**Comment by sevku** [comment]
that's certainly a good option for people who know this trick, then again, people who know this trick probably also know that they just need patience. displaying the progress of steps that take a long time certainly wouldn't be a bad idea.

**Comment by markokoskinen2037** [comment]
Same thing happened to me. I thought the program must have been bugged to "i386-pc installation progress", because it stayed there for about an hour before I aborted the process. Some kind of message that something is actually happening under the hood would be really helpful.

**Comment by Nexion** [comment]
@Lin-Buo-Ren it doesn't, unfortunately. I'm running the utility with `--verbose` flag, and after unpacking the files from iso image it just prints this and doesnt say anything for half an hour already.

```
Installing GRUB bootloader for legacy PC booting support...
Installing for i386-pc platform.
```

What is it actually doing for so long?

**Comment by jnachtigall** [comment]
I am waiting now for 2.5h for `Installing for i386-pc platform.` to complete on an 10 year old Thinkpad T400. Burning an 4GB windows 8.1 iso to a 32GB usb stick. Load is at `6.09`, cpu pretty much idle, so I think it is IO. 

FWIW, I think #251 is a duplicate of this bug here and could be closed as duplicate.

**Comment by hexathos** [comment]
Having the same problem, hangs already 20 minutes on a fresh usb stick, usb3, at 4.x 

iotop shows 99.9% on mount.ntfs process with woeusb_target

**Comment by yerbestpal** [comment]
I also have this same issue right now. I'm glad I checked here because I was ready to call it quits.

**Comment by Jpfonseca** [comment]
Using it now and I wish I knew it takes a while to finish the process before I started it.

**Comment by tycrek** [comment]
This is still an issue for me an Arch using the AUR `-git` package. Using an i7-8565U and system load is nearly idle with the GUI version.

**Comment by bluedaze** [comment]
Just going to chime in and say this definitely needs a progress indicator. Also why the heck does this step take SO long?

**Comment by bluedaze** [comment]
I was going to take a look at this to see if I could try and make some changes, but I can't seem to find the code for this? Any help would be appreciated.

**Comment by adamsb6** [comment]
It looks like this is grub2-install causing the hang. I strace'd it to find that it's sequentially copying a ton of files into the grub2 directory and is fsyncing after every write. See [here](https://gist.github.com/adamsb6/c3b8aa264aec9f5d8ff54793f1078d2b) for an example. In my case it's the fsync that's taking the most time, by far. Maybe slow fsync is a property of cheap USB flash devices.

You can't fix the problem directly in WoeUSB, but you can hack around it. You could use an LD_PRELOAD trick to make fsync a no-op. In my case that'd probably speed things up by 10X. Then to make sure that everything is properly flushed, make a sync() call after grub2-install exits.

You could also do these copies in parallel in grub2-install, but that would be outside the scope of the WoeUSB project.

**Comment by bluedaze** [comment]
> It looks like this is grub2-install causing the hang. I strace'd it to find that it's sequentially copying a ton of files into the grub2 directory and is fsyncing after every write. See [here](https://gist.github.com/adamsb6/c3b8aa264aec9f5d8ff54793f1078d2b) for an example. In my case it's the fsync that's taking the most time, by far. Maybe slow fsync is a property of cheap USB flash devices.
> 
> You can't fix the problem directly in WoeUSB, but you can hack around it. You could use an LD_PRELOAD trick to make fsync a no-op. In my case that'd probably speed things up by 10X. Then to make sure that everything is properly flushed, make a sync() call after grub2-install exits.
> 
> You could also do these copies in parallel in grub2-install, but that would be outside the scope of the WoeUSB project.

Okay, forgive me for being naive here as I have a very limited understanding of the issue but do you think that sync could be delayed all files have been written?

**Comment by adamsb6** [comment]
> Okay, forgive me for being naive here as I have a very limited understanding of the issue but do you think that sync could be delayed all files have been written?

Yep, that's what I'm suggesting. Make the fsync() calls within grub2-install a no-op, then call sync() afterwards to flush all buffers.

I tried this out, dropping page cache between runs, and found that a normal run was about 55 minutes, and the fsync no-op about 28 minutes:
regular: https://gist.github.com/adamsb6/5bdb93fe86000e2ac0975b1498846246
no-op: https://gist.github.com/adamsb6/07eaa1c704541ac7e93349f6c4c9601f

**Comment by AkgunFatih** [comment]
i will never use WoeUSB anymore.

**Comment by johnmendrake** [comment]
> i will never use WoeUSB anymore.

What is your alternative?
Had always struggles to find a good tool under Linux but if I have to wait that long, asking somebody with a Windows machine to do the job is probably faster..

**Comment by AkgunFatih** [comment]
> > i will never use WoeUSB anymore.
> 
> What is your alternative?
> Had always struggles to find a good tool under Linux but if I have to wait that long, asking somebody with a Windows machine to do the job is probably faster..

i used WoeUSB to create a Win10 media
but now now my alternative is creating it manual using wimtools by followinf this [amazing guide ](https://ypcs.fi/howto/2018/12/01/windowsiso/)

alternatively i am running win 10 in virtualbox
and then use the windows 10 media creation tool

those methods are 20x faster then woeUSB

**Comment by Lin-Buo-Ren** [comment]
@AkgunFatih 
Thanks for the info!  That's definitely can be used to improve WoeUSB etc.

**Comment by eathtespagheti** [comment]
Struggling with this issue right now, it's it possible to stop the i386 grub installation and the start a manual grub installation on the drive? Would it be faster?

**Comment by Sachsenspielt96** [comment]
@AkgunFatih But when I want to install Win in a Virtualbox, it says that Windows has no Mediadriver. Can you help me?

**Comment by AkgunFatih** [comment]
@Sachsenspielt96 add me on discord i can help you there Crystal#0702

**Comment by dualhammers** [comment]
Had the same issue with WoeUSB, and canceling the operation at the `i386-pc installation` corrupted my USB stick and now it only shows 512k space.

**Comment by olealgoritme** [comment]
it has to be the worst decision ever made, to have it 'lock' at an installation step.

**Comment by Kloppis** [comment]
Same problem here on an Dell Precision M6500 Series Notebook.
Trying to create a Windows USB Stick to install Win10 Pro on that machine.
Have been stuck at grub installation for about 15 Minutes. And thats over USB3 with a very fast flash drive.
There should be at least a message telling you that this could take a while or better a fix for that issue so its not taking that long.

**Comment by Lin-Buo-Ren** [comment]
Note that [WoeUSB 5.1.0](https://github.com/WoeUSB/WoeUSB/releases/tag/v5.1.0) support preparing installer drives from source images with >4GiB WIM archives using the FAT32 filesystem, which should eliminate most of the use cases of using NTFS as the target filesystem.

Refer [the new project's issue tracker](https://github.com/WoeUSB/WoeUSB/issues) if you have any problems with the fix, thanks!

**Comment by tz4678** [comment]
i have same problem

**Comment by tz4678** [comment]
```bash
~ 
➜ sudo woeusb --device ~/Downloads/Win10_21H1_English_x64.iso /dev/sdb --tgt-fs NTFS
[sudo] password for sergey: 
WoeUSB v3.3.1
==============================
Mounting source filesystem...
Wiping all existing partition table and filesystem signatures in /dev/sdb...
Ensure that /dev/sdb is really wiped...
Creating new partition table on /dev/sdb...
Creating target partition...
Making system realize that partition table has changed...
Wait 3 seconds for block device nodes to populate...
Cluster size has been automatically set to 4096 bytes.
Creating NTFS volume structures.
mkntfs completed successfully. Have a nice day.
--2021-09-14 00:10:41--  https://github.com/pbatard/rufus/raw/master/res/uefi/uefi-ntfs.img
SSL_INIT
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/pbatard/rufus/master/res/uefi/uefi-ntfs.img [following]
--2021-09-14 00:10:41--  https://raw.githubusercontent.com/pbatard/rufus/master/res/uefi/uefi-ntfs.img
SSL_INIT
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 524288 (512K) [application/octet-stream]
Saving to: ‘/tmp/WoeUSB.6te7gw.tempdir/uefi-ntfs.img’

uefi-ntfs.img       100%[===================>] 512.00K  --.-KB/s    in 0.1s    

2021-09-14 00:10:42 (4.27 MB/s) - ‘/tmp/WoeUSB.6te7gw.tempdir/uefi-ntfs.img’ saved [524288/524288]

1024+0 records in
1024+0 records out
524288 bytes (524 kB, 512 KiB) copied, 3.37066 s, 156 kB/s
Mounting target filesystem...
Applying workaround to prevent 64-bit systems with big primary memory from being unresponsive during copying files.
Copying files from source media...
Installing GRUB bootloader for legacy PC booting support...
Installing for i386-pc platform.

```

was tarted 1.5 hours ago
